drivers/i2c/i2c_sam_twi.c

```diff
@@ -0,0 +1,393 @@
+ */
+
+/** @file
+ * @brief I2C bus (TWI) driver for Atmel SAM MCU family.
+ *
+ * Limitations:
+ * - Only I2C Master Mode with 7 bit addressing is currently supported.
+ * - No reentrancy support.
+ */
+
+#include <errno.h>
+#include <misc/__assert.h>
+#include <stdbool.h>
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <i2c.h>
+#include "i2c-priv.h"
+
+#define SYS_LOG_DOMAIN "dev/i2c_sam_twi"
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_I2C_LEVEL
+#include <logging/sys_log.h>
+
+/** I2C bus speed [Hz] in Standard Mode */
+#define BUS_SPEED_STANDARD_HZ         100000U
+/** I2C bus speed [Hz] in Fast Mode */
+#define BUS_SPEED_FAST_HZ             400000U
+/* Maximum value of Clock Divider (CKDIV) */
+#define CKDIV_MAX                          7
+
+/* Device constant configuration parameters */
+struct i2c_sam_twi_dev_cfg {
+	Twi *regs;
+	void (*irq_config)(void);
+	u32_t bitrate;
+	const struct soc_gpio_pin *pin_list;
+	u8_t pin_list_size;
+	u8_t periph_id;
+	u8_t irq_id;
+};
+
+struct twi_msg {
+	/* Buffer containing data to read or write */
+	u8_t *buf;
+	/* Length of the buffer */
+	u32_t len;
+	/* Index of the next byte to be read/written from/to the buffer */
+	u32_t idx;
+	/* Value of TWI_SR at the end of the message */
+	u32_t twi_sr;
+	/* Transfer flags as defined in the i2c.h file */
+	u8_t flags;
+};
+
+/* Device run time data */
+struct i2c_sam_twi_dev_data {
+	struct k_sem sem;
+	struct twi_msg msg;
+};
+
+#define DEV_NAME(dev) ((dev)->config->name)
+#define DEV_CFG(dev) \
+	((const struct i2c_sam_twi_dev_cfg *const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct i2c_sam_twi_dev_data *const)(dev)->driver_data)
+
+static int i2c_clk_set(Twi *const twi, u32_t speed)
+{
+	u32_t ck_div = 0;
+	u32_t cl_div;
+	bool div_completed = false;
+
+	/*  From the datasheet "TWI Clock Waveform Generator Register"
+	 *  T_low = ( ( CLDIV × 2^CKDIV ) + 4 ) × T_MCK
+	 */
+	while (!div_completed) {
+		cl_div =   ((SOC_ATMEL_SAM_MCK_FREQ_HZ / (2 * speed)) - 4)
+			 / (1 << ck_div);
+
+		if (cl_div <= 255) {
+			div_completed = true;
+		} else {
+			ck_div++;
+		}
+	}
+
+	if (ck_div > CKDIV_MAX) {
+		SYS_LOG_ERR("Failed to configure I2C clock");
+		return -EIO;
+	}
+
+	/* Set TWI clock duty cycle to 50% */
+	twi->TWI_CWGR = TWI_CWGR_CLDIV(cl_div) | TWI_CWGR_CHDIV(cl_div)
+			| TWI_CWGR_CKDIV(ck_div);
+
+	return 0;
+}
+
+static int i2c_sam_twi_configure(struct device *dev, u32_t config)
+{
+	const struct i2c_sam_twi_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Twi *const twi = dev_cfg->regs;
+	union dev_config cfg;
+	u32_t bitrate;
+	int ret;
+
+	if (!(config & I2C_MODE_MASTER)) {
+		SYS_LOG_ERR("Master Mode is not enabled");
+		return -EIO;
+	}
+
+	if (config & I2C_ADDR_10_BITS) {
+		SYS_LOG_ERR("I2C 10-bit addressing is currently not supported");
+		SYS_LOG_ERR("Please submit a patch");
+		return -EIO;
+	}
+
+	cfg.raw = config;
+
+	/* Configure clock */
+	switch (cfg.bits.speed) {
```

Comment:
can we avoid using cfg.bits.speed and instead doing something like:

speed = (config & I2C_SPEED_MASK) >> I2C_SPEED_SHIFT;
switch (speed)